### DISCUSSION TITLE
Identifying phases in property calculations

### DISCUSSION THREAD

**Discussion by NCJazzMan65** [IDAES/idaes-pse#1528] [General]
I am creating a custom property package which contains both vapor phase and solid phase components.  I want my package to calculate the molecular weight and molar density of the vapor phase component mixture.  How do I determine which phase each component is in when I am performing these calculations so I can only include the vapor phase components?

**Comment by dallan-keylogic** [comment]
If you're creating your own property package, that's entirely up to you. The IDAES convention is to define a property `mole_frac_phase_comp` which is indexed by both phase and component sets. `mole_frac_phase_comp["Vap", j]` gives you the mole fraction of species `j` in the vapor phase.

First consider whether the vapor and solid phases should be a single property package or two separate property packages. You generally want them in a single property package if you have streams that are potentially mixed phase, like low quality steam. If each stream is a pure phase but the solid and vapor phases are brought into contact with each other in a unit operation, it might be best to have two separate property package. Then you'd have two separate control volumes, one gas and one solid, or one control volume and one state block.

See `solvent_reboiler.py`, `solvent_condenser.py`, and `solvent_column.py` [from column_models](https://github.com/IDAES/idaes-pse/tree/main/idaes/models_extra/column_models) for some options on how to set things up. Also check whether a model suitable for your purposes (or, at least, a starting point for your purposes) is present in [gas_solid_contactors](https://github.com/IDAES/idaes-pse/tree/main/idaes/models_extra/gas_solid_contactors).

**Reply by NCJazzMan65** [reply]
Thank you for your feedback.  For my situation, I need the vapor and solid phase components to be in the same property package because I'll be mixing them together into a single stream which will then go through additional unit models.  I have looked at the gas_solid_contactors models and I am using the property packages I found there as a starting point for mine.

To be a little more specific about what I'm trying to do, I'm trying to modify the following equation for mixture molecular weight (which I got from one of the gas_solid_contactors models) to only include the vapor species I've defined in my property package:

def mw_eqn(b):
      return b.mw == sum(
            b.mole_frac_comp[j] * b.params.mw_comp[j] for j in b.params.component_list
      )

You suggested using the mole_frac_phase_comp property.  Is this a property defined in the base IDAES libraries or do I need to define it?  If it exists, can I simply replace the mole_frac_comp reference in the above equation with mole_frac_phase_comp?

**Reply by dallan-keylogic** [reply]
What are your state variables, by the way?

The property you want to define is `mw_phase`, then index it by your phase set (so whatever names you gave to the vapor and solid phases. It can be defined as a `Constraint` or an `Expression`.

Properties like `mole_frac_phase_comp` are registered in `idaes.core.base.property_set` as being a standard name of a common property, but the core property package infrastructure does nothing to define them. Not every property package works on a mole basis. [See this page for related information.](https://idaes-pse.readthedocs.io/en/stable/how_to_guides/custom_models/property_package_development.html)

If you're looking for a package that defines most of those kinds of variables automatically, try out the [modular property framework.](https://idaes-pse.readthedocs.io/en/stable/explanations/components/property_package/general/index.html#modular-property-package-framework) It does support solid phases, but solid-vapor equilibrium is not defined at the moment.

Could I ask what your application is? If I knew that, it might be easier to help you.

**Reply by dallan-keylogic** [reply]
You might also be interested in [this example.](https://idaes-examples.readthedocs.io/en/latest/docs/properties/custom/custom_physical_property_packages_doc.html)

**Reply by NCJazzMan65** [reply]
I am looking to model a chemical looping process.  In case you're not familiar with this process, it involves contacting an oxygen carrier particle (ours is iron based) with different gaseous streams in a series of moving bed and fluidized bed reactors to produce either syngas or hydrogen and a almost pure stream of CO2.  We have already modeled this system in Aspen using a series of Gibbs reactors, but we would like to explore modeling this system in IDAES to take advantage of the gas-solids contactor models and the robust optimization capabilities.

We don't have all of our kinetic expressions developed yet so, as a starting point, I'm trying to duplicate a small portion of the Aspen model. Because I am using Gibbs reactors for this first model, I need all of my components in the same property package.  Further complicating things is that we have custom correlations for things like specific heat and enthalpy, which means I can't use the modular property package framework (which I have successfully used before).  So, I'm trying to figure out the custom property package approach and structure.  I have been following the example (tutorial) you pointed me to.  It's good, but there are still a few specifics the tutorial doesn't cover.

To answer the question on state variables, my variables are: mole flow, temperature, pressure, and mole fraction.

**Reply by dallan-keylogic** [reply]
First, you might be interested in the [chemical looping model in our Examples.](https://idaes-examples.readthedocs.io/en/2.6.0/docs/diagnostics/structural_singularity_doc.html). It's in the model diagnostics section because it's really easy to mess up in the system specification, and as it stands it still is badly-scaled.

The Gibbs reactor *should* support multiple phases, but it hasn't been tested.

You can still use the modular property framework with custom correlations, you just need to create functions for those correlations and provide them for things like enthalpy. Look at how the NIST and Perry correlations are given for an example of how to do it.

If you don't need the phases to be in equilibrium outside of the Gibbs reactors, your best bet is using the modular property framework with phase-component flow state variables. That way you avoid having to provide flash calculations. 

Gotta run now, if you have further questions let me know.
